### PR TITLE
Adds validation for boarding restrictions of split/joined trips

### LINKED ISSUE #33: Add validation: Split/joined trips with incorrect drop-off/pickup types
_Common train operations involve two trains that are joined at a station and then continue the journey as one train, or one train that is split at a station into two trains headed in different directions. Model joining and splitting in GTFS with two separate trips, one for each lineup of vehicles. Use pickup and drop off restrictions to prevent routing results that show duplicated trips for the shared part of the trip._
([Transit Partners Help](https://support.google.com/transitpartners/answer/7084064?hl=en))

A validation should check this.

I'll submit a PR shortly.

### PR DESCRIPTION
**Pull request by hbruch**
Any two trips which have the first or last two stops in common (same stop, same departure and arrival time on the same schedule) are compared against each other. 
For their stops in common, all but one should have NONE (1) as their `drop_off_type` (in case of split trips) or `pickup_type` (in case of joined trips).

Note: this validation assumes joined/merged trips to have the same `schedule_id`.

Fixes #33.

### CODE REVIEW

File: src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		IndexedReadOnlyDao dao = context.getDao();
+		ReportSink reportSink = context.getReportSink();
+
+		// Process trips route by route to reduce memory usage
+		// split trains have at least their two first stops in common, joined
+		// at least their last stops. We assume that both trips have the same
+		// service id. That means stop id, departure and arrival time
+		// match so we collect all split/joined trips by a key representing
+		// their first/last to stops and service id.
+		dao.getRoutes().forEach(route -> {
```
Comment (laurentg) [inline src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java]: Should not we check for route type and restrict to a (configurable?) subset of types? Joining and splitting trains make sense for rail mode, but maybe for buses this situation can be legitimate.

File: src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+						if (haveBothDropOffInSameStop(dao.getStopTimesOfTrip(tripId1),
+								dao.getStopTimesOfTrip(tripId2))) {
+							reportSink.report(
+									new WrongDropOffPickUpTypeForSplitTripsIssue(tripId1,
+											tripId2, true));
+						}
+					}
+				}
+			}
+			// validate that of split trains, at least one has dop-off none set
```
Comment (laurentg) [inline src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java]: The comment should probably says 'joined trains'?

File: src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+			// for stops in common
+			for (List<GtfsTrip.Id> joinedTripsIds : Multimaps.asMap(
```
Comment (laurentg) [inline src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java]: Should not we test also if trip headsigns for joined trains are the same?
> Set each section of a joined train to display the same destination on the trip_headsign.

**Comment by laurentg** [comment]
Excellent job, thanks! I'm currently refactoring the validator interface for stop time scanning validators to use a new kind of validator (trip and stop time streaming), to be able to optimize memory usage further and/or use external storage (see #31). I will try to merge this PR before pushing those changes, but to let you know the code will be slightly impacted.

**Comment by laurentg** [comment]
[This is the first change I'm talking about](https://github.com/mecatran/gtfsvtor/commit/89302b7c6ecc734bd582bed5370a07c8fcd1a2ba), and [the second change](https://github.com/mecatran/gtfsvtor/commit/6afd906cc92d4e5c7ce8d28051dd8cc3d543a8d0).

File: src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java (hunk 1 above)
Comment (hbruch) [reply src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java]: Indeed. I added a ConfigurableOption. Not sure if configuration should better only support primitives, but parsin route_types directly could be of value for other validators as well

File: src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java (hunk 3 above)
Comment (hbruch) [reply src/main/java/com/mecatran/gtfsvtor/validation/dao/DropOffPickUpTypeForSplitOrJoinedTripsValidator.java]: Yes, I thought about this, too. Check is added.

**Comment by laurentg** [comment]
Thanks for your code and updating to the new TripTimesValidator interface!

**Pull request merged**